### PR TITLE
[Feature] Build PPO training from an environment and networks

### PR DESCRIPTION
**Pull request by vmoens**
PR stack: `main` ← #4329 (controller deployment) ← **#4332 (this PR)** ← #4330 (MicroDuck integration and training tutorials). All branches are in `pytorch/rl`.

Companion tutorials in #4330: [MicroDuck integration](https://docs-preview.pytorch.org/pytorch/rl/4330/tutorials/microduck.html) and [low-/high-level PPO training](https://docs-preview.pytorch.org/pytorch/rl/4330/tutorials/microduck_skills.html).

Build ordinary PPO training from an environment, actor and critic. The tutorial needs the same setup twice; `PPOTrainer.from_env` moves that reusable setup into TorchRL.

```python
from torchrl.trainers.algorithms import PPOTrainer

trainer = PPOTrainer.from_env(
    env,
    actor=actor,             # Returns actions and log probabilities.
    critic=critic,           # Writes "state_value".
    total_frames=1_000_000,
    frames_per_batch=4096,
    minibatch_size=256,
)
trainer.train()
```

This constructs `Collector`, `ClipPPOLoss`, Adam, GAE, and minibatch sampling. The collector installs missing policy primers and initialization tracking. The ordinary constructor remains available for custom components.

**Recurrent policies and task-wise advantages**

```python
trainer = PPOTrainer.from_env(
    env, actor=recurrent_actor, critic=recurrent_critic,
    total_frames=10_000_000,
    frames_per_batch=4096,   # Transitions across all environments.
    minibatch_size=256,
    sub_traj_len=64,         # Consecutive steps per environment.
    gae_kwargs={"group_key": "task_id", "average_gae": True},
    loss_kwargs={"entropy_coeff": 0.01},
    num_epochs=5,
)
trainer.train()
```

Feedforward PPO samples shuffled transitions from a replay buffer. Setting `sub_traj_len` instead samples contiguous windows with `BatchSubSampler` and runs GAE in recurrent mode. When GAE normalizes within task groups, global loss normalization defaults off. Explicit loss options take precedence.

**Nested keys**

```python
# Example env specs:
# action: ("agent", "command")
# reward: ("agent", "reward")
# done:   ("agent", "done"), ("agent", "terminated")

trainer = PPOTrainer.from_env(
    env, actor=actor, critic=critic,
    value_key=("agent", "value"),
    total_frames=100_000,
)
# Action and reward keys come from env.
# Episode boundaries use the reward namespace when present, else root keys.
```

Existing `action_key`, `reward_key`, `done_key`, and `terminated_key` overrides are forwarded. Multi-agent reward and boundary tensors must have compatible shapes for GAE.

**Logging, checkpoints, and Hydra use the existing APIs**

```python
from torchrl.checkpoint import Checkpoint, GlobalRNGState

trainer = PPOTrainer.from_env(
    env, actor=actor, critic=critic, total_frames=1_000_000,
    logger=logger,
    checkpoint=Checkpoint(rng=GlobalRNGState()),
    save_trainer_file="ppo.trainer",
)
trainer.load_from_file("ppo.trainer")  # Omit for a fresh run.
trainer.train()
```

```yaml
_target_: torchrl.trainers.algorithms.PPOTrainer.from_env
total_frames: 1000000
frames_per_batch: 4096
minibatch_size: 256
sub_traj_len: 64
```

```python
from hydra.utils import instantiate

trainer = instantiate(cfg, env=env, actor=actor, critic=critic)
```

No constructor kwargs or config defaults change. Training owns and closes the collector environment; evaluation creates a fresh one.

The categorical skill-selector run also exposed a `LogScalar` regression: `std()` failed on integer actions. Convert to floating point for that statistic so categorical PPO can keep normal logging enabled.

**Validation**

- 33 trainer cases pass, covering feedforward/recurrent updates, nested keys, checkpoint resume followed by further learning, integer-action logging, sequence sampling, GAE hooks and target updates.
- The runnable factory docstring and a Hydra factory training invocation pass.
- The stacked MicroDuck tutorial executes both PPO stages with a shared GRU actor/critic, task-wise GAE, then a categorical high-level actor.
- The full launcher runs with two native MuJoCo workers; the smoke launcher saves and reloads both stages.
- Formatting, focused lint and whitespace checks pass.

### CODE REVIEW

**Comment by vmoens** [comment]
Superseded by a rewritten standalone PPO factory PR. The replacement depends only on the focused replay-sampler checkpoint fix, not on the low-level-controller feature.
